### PR TITLE
Fix optint and bit sign into x64 and x86 architectures

### PR DESCRIPTION
**Pull request by dinosaure**
/cc @cfcs 

So this branch was tested into `4.08.1+32bit` and `4.08.1` on a 64-bits processors. `optint` does not have a test (I started to make a fuzzer, but eh!), so this update is according to what `checkseum` tells me about tests (`checkseum.c` and `checkseum.ocaml`).

At the end, and as the final test, `decompress.1.0.0` was updated and tested into `4.08.1+32bit` where we obviously reach an integer overflow as expected. A patch into mirage/decompress#74 to fix it. A patch to `checkseum` will be added about `CRC32C` which has a "fast" implementation which does not work into a 32-bits architecture.

### CODE REVIEW

File: src/int_x64_backend.ml [hunk 1]
```diff
@@ -7,24 +7,36 @@ let neg x = (-x)
...
+    let q = shift_left (div (shift_right_logical n 1) d) 1 in
+    let r = sub n (mul q d) in
+    if unsigned_compare r d >= 0 then succ q else q
+
+let div a b = Nativeint.(to_int (unsigned_div (of_int a) (of_int b)))
 let rem a b = a mod b
 let succ x = x + 1
 let pred x = x - 1
-let abs x = Pervasives.abs x
+let abs x = if x >= 0 then x else (-x)
```
Comment (cfcs) [inline src/int_x64_backend.ml]:
```suggestion
let abs x =
  let mask = x asr Sys.int_size in (* extract sign: -1 if signed, 0 if not signed *)
  (x + mask) lxor mask
```
Branch-free version using [this bithacking trick](https://graphics.stanford.edu/~seander/bithacks.html#IntegerAbs).
Both implementations (current + my suggested) are the identity function when `x = min_int`.
Do you think that is better than making it be `0` or `max_int` in that case ?

Assembly for branch-free version (suggested) - I also checked the other variant they have (with xor first, then minus; that resulted in one extra instruction):
```asm
        movq    %rax, %rbx
        sarq    $63, %rbx
        orq     $1, %rbx
        leaq    -1(%rax,%rbx), %rax
        xorq    %rbx, %rax
        orq     $1, %rax
        ret
```

Asm for branching version (current):
```asm
        cmpq    $1, %rax
        jl      .L102
        ret
.L102:
        movq    $2, %rbx
        subq    %rax, %rbx
        movq    %rbx, %rax
        ret
```

File: src/int_x64_backend.ml [hunk 2]
```diff
@@ -7,24 +7,36 @@ let neg x = (-x)
...
 let max_int = Int32.(to_int max_int)
 let min_int = Int32.(to_int min_int)
 let logand a b = a land b
 let logor a b = a lor b
 let logxor a b = a lxor b
-let lognot x = lnot (x land 0xffffffff)
+let lognot x = lnot x
 let shift_left a n = a lsl n
 let shift_right a n = a lsr n
 let shift_right_logical a n = a asr n
```
Comment (cfcs) [inline src/int_x64_backend.ml]: `shift_right_logical` should be `lsr` (logical shift right) and `shift_right` should be `asr` (arithmetic shift right); they're switched here.

**Review by cfcs — CHANGES_REQUESTED** [review]

File: src/int_x64_backend.ml [hunk 3]
```diff
@@ -33,6 +45,23 @@ let to_string x = string_of_int x
...
 
+let bit_sign = 0x80000000
+let without_bit_sign (x:int32) = if x >= 0l then x else Int32.logand x (Int32.lognot 0x80000000l)
+
+let invalid_arg fmt = Format.kasprintf invalid_arg fmt
+
+let to_int32 x =
+  if x land (lnot 0xffffffff) <> 0
+  then invalid_arg "Optint.to_int32: %d can not fit into a 32 bits integer" x
+  else Int32.of_int (x land 0xffffffff)
```
Comment (cfcs) [inline src/int_x64_backend.ml]: Here it's maybe faster to do
```ocaml
  let truncated = x land 0xff_ff_ff_ff in
  if truncated <> x
  then invalid_arg "Optint.to_int32: %d can not fit into a 32 bits integer" x
  else Int32.of_int truncated
```

File: src/int_x64_backend.ml (hunk 1 above)
Comment (dinosaure) [reply src/int_x64_backend.ml]: I would like to keep same behaviors than OCaml:
```ocaml
# abs min_int
- : int = -4611686018427387904
# Int32.abs Int32.min_int
- : int32 = -2147483648l
```

But I like your trick 👍

**Comment by dinosaure** [comment]
Ok let's start to merge and release.

**Pull request merged**